functions.php

```diff
@@ -0,0 +1,27 @@
+<?php
+
+/**
+ * @params $db - a database object returned from the PDO at the top of the page
+ *
+ * @returns - an assoc array containing all of the data from the database
+ **/
+function getAllFromDB(object $db): array
+{
+    $query = $db->prepare('SELECT * FROM `books`;');
+    $query->execute();
+    return $query->fetchAll();
+}
+
+/**
+ * @param array $query - generated from getAllFromDB()
+ *
+ * @return array - an array filled with HTML strings ready to be output
+ */
+function prepareOutput(array $query): array
+{
+    $result = [];
```

Comment:
I do like this approach of constructing an array with the output for each item. But one thing to bear in mind is that it results in having to iterate through the data two times, once at this point and then again when it gets displayed.

This is almost going into the realms of premature optimization, so not really saying you have to change it. Just something to think about!

if you were more worried about performance it would be best to instead concat each line into a string and return that string - that way there only needs to be one loop